### ISSUE TITLE
Facet needs to instantiate empty groups

### ISSUE THREAD

**Issue by domoritz** [vega/vega#781]
When facetting data and grouping by multiple dimensions (with cardinalities m and n), we expect n by m groups being created. However, if there are no rows for a particular partition of data, we don't get a group. To support this case, we need an option in facet to generate groups for empty data (similar to https://vega.github.io/vega/docs/transforms/aggregate/, I believe).

**Comment by kanitw** [comment]
To clarify, this is only important when we facet by multiple fields. 

For example, when we facet by both row and column like this VL spec (and its output [Vg spec](https://pastebin.com/burQYKGb)), the number of groups does not match # row x # column due to missing groups.
```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "data/cars.json"},
  "mark": "point",
  "encoding": {
    "row": {"field": "Cylinders","type": "ordinal"},
    "column": {"field": "Origin","type": "ordinal"},
    "x": {"field": "Horsepower","type": "quantitative"},
    "y": {"field": "Miles_per_Gallon","type": "quantitative"}
  }
}
```

**Comment by jheer** [comment]
Top-level group items for facets are determined via an aggregate operation, so that is likely the proper place to pursue this. However, there is currently no mechanism for an aggregate to "hallucinate" unobserved data. The current parameters only allow groups (initially with non-zero item counts) to be kept around if the count drops to zero due to subsequent filtering operations.

In other words, aggregate performs nesting and only nesting. It seems that what you are requesting is instead an aggregation domain that is the cross-product of two "univariate" aggregates. Is that correct? If so, there may be multiple ways to achieve this and I'll need to think about which way we should prefer.

**Comment by domoritz** [comment]
> It seems that what you are requesting is instead an aggregation domain that is the cross-product of two "univariate" aggregates.

Yes, we need the cross product of the distinct items of the group by dimensions. 

For example if we have a data source

```
id,a,b
1,1,3
2,2,3
3,2,4
4,3,4
5,3,4
```

and facet it by `a` and `b`. 

We need to group the tuples like this

```
group, tuples
(1,3), 1
(1,4), no tuples but we need a group
(2,3), 2
(2,4), 3
(3,3), no tuples but we need a group
(3,4), 4 and 5
```

**Comment by jheer** [comment]
Added with updates to vega-dataflow and vega-parser. Will be included in version v3.0.0-beta.30.
